src/mono/browser/runtime/jiterpreter-trace-generator.ts

```diff
@@ -3837,102 +3837,107 @@ function emit_simd_3 (builder: WasmBuilder, ip: MintOpcodePtr, index: SimdIntrin
         builder.appendSimd(WasmSimdOpcode.v128_const);
-        builder.appendBytes(newShuffleVector);
+        builder.appendBytes(nativeIndices);
     } else {
         // Load indices (in chars)
         append_ldloc(builder, indicesOffset, WasmOpcode.PREFIX_simd, WasmSimdOpcode.v128_load);
-        // There's no direct narrowing opcode for i32 -> i8, so we have to do two steps :(
-        if (elementCount === 4) {
-            // i32{lane0 ... lane3} -> i16{lane0 ... lane3, 0 ...}
-            builder.v128_const(0);
-            builder.appendSimd(WasmSimdOpcode.i16x8_narrow_i32x4_u);
-        }
-        // Load a zero vector (narrow takes two vectors)
-        builder.v128_const(0);
-        // i16{lane0 ... lane7} -> i8{lane0 ... lane7, 0 ...}
-        builder.appendSimd(WasmSimdOpcode.i8x16_narrow_i16x8_u);
-        // i8{0, 1, 2, 3 ...} -> i8{0, 0, 1, 1, 2, 2, 3, 3 ...}
-        builder.appendSimd(WasmSimdOpcode.v128_const);
-        for (let i = 0; i < elementCount; i++) {
-            for (let j = 0; j < elementSize; j++)
-                builder.appendU8(i);
-        }
-        builder.appendSimd(WasmSimdOpcode.i8x16_swizzle);
-        // multiply indices by 2 or 4 to scale from elt indices to byte indices
-        builder.i32_const(elementCount === 4 ? 2 : 1);
-        builder.appendSimd(WasmSimdOpcode.i8x16_shl);
-        // now add an offset to the additional bytes of each lane, i.e.
-        // 0 1 2 3 0 1 2 3 ...
-        builder.appendSimd(WasmSimdOpcode.v128_const);
-        for (let i = 0; i < elementCount; i++) {
-            for (let j = 0; j < elementSize; j++)
-                builder.appendU8(j);
+        if (elementCount !== 16) {
+            const shift = elementCount === 8 ? 1 : elementCount === 4 ? 2 : 3;
+
+            if (checkBounds) {
+                builder.local("shuffle_indices", WasmOpcode.tee_local);
+
+                builder.appendSimd(WasmSimdOpcode.v128_const);
+                for (let i = 0; i < elementCount; i++) {
+                    builder.appendU8(elementCount - 1);
+                }
+                if (elementSize === 2)
+                    builder.appendSimd(WasmSimdOpcode.i16x8_gt_u);
+                else if (elementSize === 4)
+                    builder.appendSimd(WasmSimdOpcode.i32x4_gt_u);
+                else if (elementSize === 8) {
+                    builder.appendSimd(WasmSimdOpcode.i64x2_gt_s);
+                }
+
+                builder.local("shuffle_indices");
+                builder.appendSimd(WasmSimdOpcode.v128_or);
+            }
+
+            builder.i32_const(shift);
+            builder.appendSimd(WasmSimdOpcode.i8x16_shl);
+
+            builder.appendSimd(WasmSimdOpcode.v128_const);
+            for (let i = 0; i < elementCount; i++) {
+                for (let j = 0; j < elementSize; j++)
+                    builder.appendU8(i * elementSize);
+            }
+            builder.appendSimd(WasmSimdOpcode.i8x16_swizzle);
```

Comment:
We should add a comment explaining what this swizzle is for, it looks wrong to me at a glance but it was there before

Reply:
I think this swizzle was originally to replicate the lanes after the narrowing conversions; you've removed the narrowing conversions but I think it's probably still replicating the lanes in a way that is desirable

Reply:
i.e. for i32 indices you'll have 3 zero bytes and an index byte, so this replicates the index byte across the zero bytes based on the constant vector that we just pushed (seems right)

Reply:
I've added comments